Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp

```diff
@@ -49,48 +50,73 @@ static ArrayBuffer* array_buffer_object_from(VM& vm, GlobalObject& global_object
 // 25.1.5.3 ArrayBuffer.prototype.slice, https://tc39.es/ecma262/#sec-arraybuffer.prototype.slice
 JS_DEFINE_NATIVE_FUNCTION(ArrayBufferPrototype::slice)
 {
-    const auto start = vm.argument(0);
-    const auto end = vm.argument(1);
-
     auto array_buffer_object = array_buffer_object_from(vm, global_object);
     if (!array_buffer_object)
         return {};
 
     // FIXME: Check for shared buffer
     // FIXME: Check for detached buffer
 
-    const auto len = array_buffer_object->byte_length();
+    auto length = array_buffer_object->byte_length();
 
-    const auto relative_start = start.is_negative_infinity()
-        ? 0
-        : start.to_integer_or_infinity(global_object);
+    auto relative_start = vm.argument(0).to_integer_or_infinity(global_object);
     if (vm.exception())
         return {};
 
-    const auto first = relative_start < 0
-        ? max(len + relative_start, 0.0)
-        : min(relative_start, static_cast<double>(len));
+    double first;
+    if (relative_start < 0)
+        first = max(length + relative_start, 0.0);
+    else
+        first = min(relative_start, (double)length);
 
-    const auto relative_end = end.is_undefined()
-        ? len
-        : end.to_integer_or_infinity(global_object);
+    auto relative_end = vm.argument(1).is_undefined() ? length : vm.argument(1).to_integer_or_infinity(global_object);
     if (vm.exception())
         return {};
 
-    double final_;
-    if (end.is_negative_infinity())
-        final_ = 0;
-    else if (relative_end < 0)
-        final_ = max(len + relative_end, 0.0);
+    double final;
+    if (relative_end < 0)
+        final = max(length + relative_end, 0.0);
     else
-        final_ = min(relative_end, static_cast<double>(len));
+        final = min(relative_end, (double)length);
+
+    auto new_length = max(final - first, 0.0);
+
+    auto constructor = species_constructor(global_object, *array_buffer_object, *global_object.array_buffer_constructor());
+    if (vm.exception())
+        return {};
 
-    const auto new_len = max(final_ - first, 0.0);
+    MarkedValueList arguments(vm.heap());
+    arguments.append(Value(new_length));
+    auto new_array_buffer = vm.construct(static_cast<Function&>(*constructor), static_cast<Function&>(*constructor), move(arguments), global_object);
```

Comment:
This is a bit ugly with those casts... if the object returned by `species_constructor()` is always a function, can we just let it return a `Function*` right away?